API, Core, Spark 3.5: Parallelize reading of deletes and cache them on executors

aokolnychyi opened the pull request:
This PR has code to parallelize reading of deletes and enable caching them on executors.

I also have a follow-up change to assign tasks for one partition to the same executor, similar to `KafkaRDD`. There is no way to express task affinity so we can only rely on task locality. The solution in `KafkaRDD` is simple to implement but won't work well if dynamic allocation is enabled (so it should be hidden under a flag).

More thoughts in [this](https://docs.google.com/document/d/1M4L6o-qnGRwGhbhkW8BnravoTwvCrJV8VvzVQDRJO5I) doc.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/SystemConfigs.java:
```diff
@@ -42,6 +42,13 @@ private SystemConfigs() {}
+      new ConfigEntry<>(
+          "iceberg.worker.delete-num-threads",
+          "ICEBERG_WORKER_DELETE_NUM_THREADS",
+          4 * Runtime.getRuntime().availableProcessors(),
```
This value may sound ridiculous but here is my thought process: there is one such thread pool per JVM, each core in an executor can get a data task that may need to load 1 to many delete files, these tasks are I/O intensive. This value essentially means we can try to load 4 delete files concurrently per each data task.

aokolnychyi replied:
Yeah, it is configurable, we just need to make sure the default value is reasonable.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/util/CharSequenceMap.java:
```diff
@@ -0,0 +1,208 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+public class CharSequenceMap<T> implements Map<CharSequence, T>, Serializable {
```
Similar to `CharSequenceSet`.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -125,6 +126,25 @@ public static StructLikeSet toEqualitySet(
     }
   }
 
+  public static <T extends StructLike> CharSequenceMap<PositionDeleteIndex> toPositionIndexes(
```
Unlike `toPositionIndex` below, this one does not filter the deletes for a particular data file. Instead, it builds an index for each referenced data file and returns a map.

aokolnychyi replied:
Added.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/util/ThreadPools.java:
```diff
@@ -59,6 +65,20 @@ public static ExecutorService getWorkerPool() {
+   * @return an {@link ExecutorService} that uses the delete worker pool
+   * @see SystemConfigs#DELETE_WORKER_THREAD_POOL_SIZE
+   */
+  public static ExecutorService getDeleteWorkerPool() {
```
We don't usually use `get` but we have `getWorkerPool()` above so I matched it for consistency.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,209 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BaseDeleteLoader implements DeleteLoader {
```
I added this abstraction cause it was too much to put into `DeleteFilter`. Take a look at the parent interface.

aokolnychyi replied:
I questioned this a bit and do believe it is better to have this logic separately from `DeleteFilter`. That said, the exact API and whether it should be a top-level class are still open questions.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,209 @@
+public class BaseDeleteLoader implements DeleteLoader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BaseDeleteLoader.class);
+  private static final long MAX_CACHE_FILE_SIZE = 32 * 1024 * 1024; // 32 MB
```
Configurable?

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -243,20 +236,9 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
-      return createDeleteIterable(records, isDeleted);
-    }
-
-    return hasIsDeletedColumn
```
This is a separate discussion but I think we should drop streaming position deletes. We originally added them when we used a set, not bitmaps.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/DeleteLoader.java:
```diff
@@ -0,0 +1,31 @@
+import org.apache.iceberg.deletes.PositionDeleteIndex;
+import org.apache.iceberg.util.StructLikeSet;
+
+public interface DeleteLoader {
```
Missing docs, not sure we would keep this abstraction. The goal is to make `DeleteFilter` simpler and focus on filtering, rather than loading and caching.

singhpk234 commented:
Should we apply some intelligence on how we are distributing the taskGroups so that we could utilize the max from the executor cache ? For ex : lets say we could prefer sending those set of data files which have a lot of overlapping delete files or may be belong to some partition (for ex : position deletes) ?

aokolnychyi commented:
> Should we apply some intelligence on how we are distributing the tasks so that we could utilize the max from the executor cache ? For ex : lets say we could prefer sending those set of data files which have a lot of overlapping delete files or may be belong to some partition (for ex : position deletes) ?

@singhpk234, I have a follow-up change to do that. Unfortunately, it is a bit controversial. There is no way to express task affinity in Spark, only locality. The best option for us is to implement what `KafkaRDD` does. The problem is that it only works well if dynamic allocation is disabled. Even without that, this feature should be useful.

singhpk234 commented:
Thanks @aokolnychyi looking forward to it :) !

aokolnychyi commented:
I tested this PR on a cluster a bit. It would be nice if someone could also play around with it in their environment.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -452,6 +454,59 @@ private static void checkSchemaCompatibility(
     }
   }
 
+  public static long defaultSize(Types.NestedField field) {
```
I think we have 2 options to limit the size of the cache:
- Add ways to limit the size of equality and position delete files to be loaded.
- Add ways to indicate the maximum cache size in bytes.

The first one is easy but we need to account that delete files may store extra columns and it will be hard to configure these options correctly cause users have no idea how much the cached representation will actually occupy. That's why I went with the estimation approach and let users configure the max cache size in bytes.

RussellSpitzer replied:
Why do we need to cache the extra columns? We wouldn't be using them?

aokolnychyi replied:
The idea is to ask the user for the total size of the cache in memory and estimate the actual size to honor those values. For equality deletes, we will rely on the number of records and types of equality columns.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/deletes/PositionDeleteIndex.java:
```diff
@@ -44,4 +44,14 @@ public interface PositionDeleteIndex {
   /** Returns true if this collection contains no element. */
   boolean isEmpty();
+
+  /** Returns true if this collection contains elements */
```
Missed a dot here.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/SystemConfigs.java:
```diff
@@ -43,14 +43,14 @@ private SystemConfigs() {}
           "iceberg.worker.delete-num-threads",
           "ICEBERG_WORKER_DELETE_NUM_THREADS",
-          Math.max(2, Runtime.getRuntime().availableProcessors()),
+          Math.max(2, 4 * Runtime.getRuntime().availableProcessors()),
```
Not a huge deal but we are avoiding the RevCheck here by putting our multiplier in a constant here. We should probably move the 4 into a field so future modifications trigger the Rev checker

aokolnychyi replied:
I can do that but is it something we want to expose to others? Would the goal be to bring the attention or to prohibit future modifications?

RussellSpitzer replied:
You can keep it private, it's to prohibit future mods

aokolnychyi replied:
Looks like we don't do that for other properties as well.
I'd be open to explore that but for all properties in a separate PR.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/BitmapPositionDeleteIndex.java:
```diff
@@ -27,6 +27,15 @@ class BitmapPositionDeleteIndex implements PositionDeleteIndex {
     roaring64Bitmap = new Roaring64Bitmap();
   }
 
+  void merge(PositionDeleteIndex other) {
```
Why not just only allow BitmapPositionDeleteIndex here? Do we not have the type when we call merge?

aokolnychyi replied:
There is also `EmptyPositionDeleteIndex`. I actually started with `BitmapPositionDeleteIndex`. I may need to go back and check with fresh eyes.

aokolnychyi replied:
I switched to accepting `BitmapPositionDeleteIndex` here and moved casts to the utility class.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/EmptyPositionDeleteIndex.java:
```diff
@@ -0,0 +1,69 @@
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
```
Since this is a static class isn't this enough? return this == other?

aokolnychyi replied:
I guess it may be serialized and deserialized to get another instance of the class.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -500,4 +501,21 @@ public static Snapshot latestSnapshot(TableMetadata metadata, String branch) {
+      return false;
+    }
+
+    String value = snapshot.summary().get(SnapshotSummary.TOTAL_DELETE_FILES_PROP);
```
Should we note here this is not a spec authorized field. A client is not required to put in TOTAL_DELET_FILES_PROP with the actual total number of delete files (although it would be weird if they did)

aokolnychyi replied:
That's a good idea, I'll add a note.

aokolnychyi replied:
Added.

aokolnychyi commented:
The Flink test failure does not seem related.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/util/ThreadPools.java:
```diff
@@ -68,8 +68,9 @@ public static ExecutorService getWorkerPool() {
-   * <p>The size of the delete worker pool limits the number of threads used to compute the
-   * PositionDeleteIndex from the position deletes for a data file.
+   * <p>The size of this worker pool limits the number of tasks concurrently reading delete files
+   * within a single JVM. In most cases, deletes are loaded while reading data on executors. The
```
The second sentence is not super useful since we have no guidance on what that size should be. I would probably just drop it, or replace it with "increase this if your executors are  larger?"

In general the whole comment is also very Spark specific ... Maybe just keep it as simple as possible for now? I don't have strong feelings here, except for removing the "has to be big enough" sentence.

aokolnychyi replied:
Agreed, will make it more generic.

aokolnychyi replied:
Made it more generic and added a note about sharing the pool between threads.

aokolnychyi commented:
@RussellSpitzer @szehon-ho, could you take a look? I am adding tests and benchmarks but the main code is ready for another round.

szehon-ho commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,260 @@
+   *
+   * <p>This method will be called only if {@link #canCache(long)} returned true.
+   */
+  protected <V> V get(String key, Supplier<V> valueSupplier, long valueSize) {
```
get() is a bit confusing, can we call it getOrLoad (like on the calling method)?

aokolnychyi replied:
Makes sense, I'll rename to `getOrLoad`.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  private void storeMapping(String executionId, String key) {
+    Collection<String> keys =
```
optional:  is there any need to define variable keys, can we just inline the whole statement:

keysByExecutionId
   .computeIfAbsent(executionId, id -> Queues.newConcurrentLinkedQueue())
   .add(key)

aokolnychyi replied:
Applied.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+    return MAX_ENTRY_SIZE;
+  }
+
+  public <V> V get(String executionId, String key, Supplier<V> valueSupplier, long valueSize) {
```
can we add javadocs to this class methods?

aokolnychyi replied:
Will do.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+      return valueSupplier.get();
+    }
+
+    storeMapping(executionId, key);
```
Sorry im confused here (will need to take another look with fresh eyes).  How does this guarantee uniqueness of value for given executionId and key?  (we store key with executionId, but then store the key in cache() anyway, regardless of whether key was used in another executionId?)

Why not just use executionId and key as cache key, or have a map of maps?

szehon-ho replied:
Comment still looks unresolved?

aokolnychyi replied:
I wanted to make keys accessible across execution IDs but I don't think it is true anymore. Let me take a look with fresh eyes now.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  public static void cleanUp(String executionId) {
+    if (instance != null) {
```
Should we throw an error if instance == null and cleanUp gets called?

aokolnychyi replied:
I was thinking we would always call cleanUp once the execution ID is done. If there is no cache, we would just do nothing. Does that make sense? Otherwise, we will have to check if we cached anything.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  public <V> V get(String executionId, String key, Supplier<V> valueSupplier, long valueSize) {
+    if (valueSize > MAX_ENTRY_SIZE) {
```
I'm a little confused here, if I try to get the cache here for this executionId and key, and it already is set in the cache I may still error out.

For example

get(id_1, key_1) - with 64mb Cache Entry // Inits cache
get(id_1, key_1) - with same 64mb Cache entry // Fails although value is already set

aokolnychyi replied:
I am not sure I got. You mean calling this method with different value estimates?

aokolnychyi commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -452,6 +454,68 @@ private static void checkSchemaCompatibility(
+        return 8;
+      case STRING:
+        // 12 (header) + 12 (fields) + 16 (array overhead) + 20 (10 chars, 2 bytes each) = 60 bytes
+        return 60;
```
I think you are right that the number of fields has changed. Instead of 3 int variables, we have 1 boolean, 1 byte, 1 int in newer versions of Java. I'll update.

szehon-ho replied:
OK, also my feeling that estimating average string sizes will be tricky given the data is quite variable, this might lead to cascading errors in estimating whether to cache or not?   Not sure any better idea, do you think its worth to make a config for it and document it in javadoc?

aokolnychyi replied:
Synced a bit offline on this one. I think it should be fine to start with this and then see actual use cases when it will become a problem. There are multiple ways to approach it but all of them are substantially more complicated.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -452,6 +454,68 @@ private static void checkSchemaCompatibility(
+   * @param field a field for which to estimate the size
+   * @return the estimated size in bytes of the field's value in memory
+   */
+  public static long defaultSize(Types.NestedField field) {
```
I was inspired by Spark but I agree `estimateSize` is a better option.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/deletes/EmptyPositionDeleteIndex.java:
```diff
@@ -0,0 +1,55 @@
+
+  @Override
+  public String toString() {
+    return "PositionDeleteIndex{}";
```
Not sure. Is `{}` not clear? I was following what we did in `EmptyStructLike`. I am flexible here.

szehon-ho replied:
Ah ok, I think i did not see {} here.  But actually, realize that we dont have toString on the other PositionDeleteIndex implementations, so didnt realize that those print with actual content, maybe we should add at some point?

aokolnychyi replied:
Agree, that would be nice to have a reasonable `toString()` implementation in others.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/deletes/PositionDeleteIndex.java:
```diff
@@ -44,4 +44,14 @@ public interface PositionDeleteIndex {
   boolean isEmpty();
+
+  /** Returns true if this collection contains elements. */
+  default boolean isNotEmpty() {
```
I added this to avoid negation inside `if` statements. In my view, those are harder to read. Scala, for instance, has helper methods for both in collections.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,260 @@
+    }
+  }
+
+  private Iterable<StructLike> loadEqDeletes(DeleteFile deleteFile, Schema projection) {
```
I was differentiating by using shorter names but let me think a bit more about this.

szehon-ho replied:
Comment not addressed yet?  I still think adding 'internal' is more clear than just shortening.  Or maybe loadEqDeletesFromFile

aokolnychyi replied:
I switched to `readEqDeletes` because Spotless started to format some lines in a weird way.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,260 @@
+    return CloseableIterable.transform(records, wrapper::copyFor);
+  }
+
+  private <T> Iterable<T> materialize(CloseableIterable<T> iterable) {
```
This essentially materializes the iterable and closes the resources so that the actual value can be cached. We cannot cache the iterable directly cause it is lazy. Let me think.

aokolnychyi replied:
Added a comment for now. Still thinking.

aokolnychyi replied:
I'd probably stick with the comment.

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -224,14 +223,10 @@ public Predicate<T> eqDeletedRowFilter() {
   public PositionDeleteIndex deletedRowPositions() {
-    if (posDeletes.isEmpty()) {
-      return null;
+    if (deleteRowPositions == null && !posDeletes.isEmpty()) {
```
We did not have it before so I think it should be OK. Also, adding double check locking would require an extra field to synchronize as `deleteLoader()` already synchronizes on `this`.

Let me think a bit about this.

aokolnychyi replied:
I think it should be OK for now given that we didn't have it before.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -452,6 +454,68 @@ private static void checkSchemaCompatibility(
+        return 60;
+      case UUID:
+        // 12 (header) + 16 (two long variables) = 28 bytes
+        return 28;
```
should we express these as OBJECT_HEADER

aokolnychyi replied:
Do you mean like an actual expression?

szehon-ho replied:
I mean like OBJECT_HEADER + 16?

szehon-ho replied:
Comment is still not addressed?

aokolnychyi replied:
After thinking a bit more, I am not sure it would make it any better. The second number would still be some abstract number. I feel it should be OK to just explain how the number was derived and use the constant for nested types. We have real computations there.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -452,6 +454,68 @@ private static void checkSchemaCompatibility(
+        long entrySize = OBJECT_HEADER + defaultSize(map.keyType()) + defaultSize(map.valueType());
+        return OBJECT_HEADER + 5 * entrySize;
+      default:
+        return 16;
```
why 16?

aokolnychyi replied:
Just a reasonable default value in case we did not cover or add another type? Open to ideas.

szehon-ho replied:
I feel typically we throw execptions in this case?

aokolnychyi replied:
I don't think we should fail queries in such cases. It is just an estimate so we better use a non-precise estimate rather than fail queries.

aokolnychyi replied:
I believe those places are a bit different. For instance, if we can't convert a Spark value, it is a correctness issue. Here, it is an optimization that is not strictly required to read or write a dataset.

szehon-ho replied:
Agree its not in the critical path.  I was more thinking in the lines of , if i ever add a new type, how will i know to add a estimate here?  Hopefully here a unit test will fail, and we will soon add a proper estimate.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,197 @@
+    return instance;
+  }
+
+  public static void cleanUp(String executionId) {
```
It seems "instance" itself will not be cleaned up (nullified), is it an issue?

aokolnychyi replied:
I don't think so cause the cache will be live as long the executor JVM is running. It just may not hold anything.

szehon-ho replied:
I was just wondering, is there any place we know we can safely GC the entire instance?  Yea I guess it will not be huge, but just for completeness

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:
```diff
@@ -70,4 +72,18 @@ private SparkSQLProperties() {}
+      "spark.sql.iceberg.executor-cache.max-entry-size";
+  public static final long EXECUTOR_CACHE_MAX_ENTRY_SIZE_DEFAULT = 64 * 1024 * 1024; // 64 MB
+
+  public static final String EXECUTOR_CACHE_MAX_TOTAL_SIZE =
```
We should document these confs

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:
```diff
@@ -70,4 +72,18 @@ private SparkSQLProperties() {}
+  public static final String EXECUTOR_CACHE_TIMEOUT = "spark.sql.iceberg.executor-cache.timeout";
+  public static final Duration EXECUTOR_CACHE_TIMEOUT_DEFAULT = Duration.ofMinutes(10);
+
+  public static final String EXECUTOR_CACHE_MAX_ENTRY_SIZE =
```
what is the guidance for max entry size?  Are we trying to prevent one huge entry preventing smaller entries from getting cached, assuming that we may cache fewer total deletes that way?

aokolnychyi replied:
Yep, that's the use case.

aokolnychyi commented:
@singhpk234 @RussellSpitzer @szehon-ho, I rebased this. I addressed most comments, I am working on tests and docs. There are a few open questions too. I'll take a look at them tomorrow.

szehon-ho reviewed (commented):
Thanks, looks better, looks like there's still some unresolved comments

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java:
```diff
@@ -0,0 +1,260 @@
+    }
+  }
+
+  private CharSequenceMap<PositionDeleteIndex> loadPosDeletes(DeleteFile deleteFile) {
```
Also renamed.

aokolnychyi commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkExecutorCache.java:
```diff
@@ -0,0 +1,228 @@
+      long start = System.currentTimeMillis();
+      V value = valueSupplier.get();
+      long end = System.currentTimeMillis();
+      LOG.info("Loaded value for {} with size {} in {} ms", key, valueSize, (end - start));
```
I am going back and forth on the log level here. I'd say it is a fragile place and it is better to always have more logs for now. I don't expect a huge number of these lines. We do have pretty detailed logs for broadcasts in Spark, for instance. 

That said, we can switch to debug if everyone thinks it would be better.

szehon-ho reviewed (approved):
Left one comment, but rest looks good to me now.  Thanks for the changes

aokolnychyi commented:
I gave this PR a round of testing on the cluster and it seems to work as expected.

aokolnychyi commented:
Thanks for reviewing, @szehon-ho @RussellSpitzer!
